Install script's chsh not working on Fedora

ptamarit opened the pull request:
Running the install script on Fedora 20 (after having yum-installed zsh) throws the following error:

    chsh: "/usr/bin/zsh" is not listed in /etc/shells

This is caused by the following line:

    chsh -s `which zsh`

The explaination is the following:
* Running `which -a zsh` return `/usr/bin/zsh` and `/bin/zsh`, but running `which zsh` returns only the first one: `/usr/bin/zsh`.
* Checking the available shells with `chsh -l | grep /zsh$` only returns `/bin/zsh`
* That's why the `chsh -s` command fails.

Both zsh paths should be present in `/etc/shells`, but it looks like it's a Fedora bug (see [bug 1034060](https://bugzilla.redhat.com/show_bug.cgi?id=1034060)).

One way to make the shell switching more robust is to set zsh from the list of available shells returned by `chsh -l` instead. The proposed patch gets the first available shell matching `/zsh$` and sets it as the current shell.

Fixes #3026, fixes #3779, fixes #3780

mcornella commented:
This is ready to merge, but it needs a rebase. Ping @ptamarit 
/cc @robbyrussell

ptamarit commented:
@mcornella The PR is now rebased. In the meantime, a check had been added to call `chsh` only if zsh is not the current shell. I changed this test as well, so we rely solely on `/etc/shells`.
/cc @robbyrussell

mcornella commented:
I disagree with the check change, we need to make sure that the current shell is not a zsh. 
Our `grep...` command outputs "the last zsh available in `/etc/shells`", but just because that's the best "heuristic" we can make. A better check would be looking if `$SHELL` ends in `zsh`: 
`${SHELL:t} = "zsh"`. I haven't had time to test it though

apjanke commented:
While we're working on this, can we use the same logic to locate `bash` in `uninstall.sh` where it's reverting your shell? It's currently using a hardcoded `/bin/bash`, which doesn't exist on all systems. E.g. in FreeBSD 10.1 it's `/usr/local/bin/bash`.
```
diff --git a/tools/uninstall.sh b/tools/uninstall.sh
index 41d6015..c9cf3e1 100644
--- a/tools/uninstall.sh
+++ b/tools/uninstall.sh
@@ -21,7 +21,8 @@ then
   source ~/.zshrc;
 else
   echo "Switching back to bash"
-  chsh -s /bin/bash
+  BASH=$(grep /bash$ /etc/shells | tail -1)
+  chsh -s $BASH
   source /etc/profile
 fi
```

Could go a little further and add a test to see if `bash` is even installed. It's not installed by default on some systems.

mcornella commented on tools/install.sh:
```diff
@@ -32,9 +32,9 @@ sed -i -e "/export PATH=/ c\\
 " ~/.zshrc
 
-if [ "$SHELL" != "$(which zsh)" ]; then
+if [ "$SHELL" != "$(grep /zsh$ /etc/shells | tail -1)" ]; then
```
Can you change this to 
```sh
if [ "${SHELL:t}" != "zsh" ]; then
```
as it is now, if `$SHELL` is zsh but not the last one in `/etc/shells` it will trigger a `chsh`.

apjanke replied:
I don't think `${PARAM:t}` works in basic `sh`, which `install.sh` runs under. How about `CURRENT_SHELL=$(expr "$SHELL" : '.*/\(.*\)'); if [ "$CURRENT_SHELL" != "zsh" ]; then ...`?

mcornella replied:
Good point, we should take this chance to put zsh as the default runner and implement the check for installed zsh that so many people have asked already.

ncanceill replied:
Agreed. This line should be:
```zsh
if [ $(expr "$SHELL" : '.*/\(.*\)') != "zsh" ]; then
```
And with that I think we are good to go.

ncanceill replied:
I submitted [ptamarit/oh-my-zsh#1](https://github.com/ptamarit/oh-my-zsh/pull/1) to deal with that.

ncanceill commented:
@ptamarit can you add "Fixes #3026" in the description text of the PR, so that this related issue gets auto-closed at merge time?

ncanceill commented:
This would also fix #3779 and #3780.

ptamarit commented:
Sorry for having been sleeping on this one.
I merged @ncanceill's pull request and rebased with master, so I guess it's ready for prime time.

ncanceill commented:
Nice! Can you squash my commit into yours?

ptamarit commented:
@ncanceill Done, I smashed your commit into mine.

mcornella commented:
This will be merged soon. Please update the body of the PR to auto-close other duplicates of this same PR: add this "Fixes #3780, fixes #3779, fixes #3799".
Also, #3799 appears to be adding logic to inform the user when zsh is not installed. That will be dealt with in a separate PR, I hope.
Thanks

ptamarit commented:
@mcornella I edited the body of the PR. The commit already contained the list of issues to auto-close so I guess it's a bit redundant...

mcornella commented:
:sweat_smile: Oops! I forgot to look there.. Anyway this is ready to merge
:+1:

ncanceill commented:
Excellent!
